File: python/examples/02_open_close_loop.py
"""Example 2: Repeatedly open and close the gripper.

Usage:
    python 02_open_close_loop.py /dev/ttyUSB0 [cycles]
    python 02_open_close_loop.py COM3 [cycles]
"""
import signal
import sys

from aidin_gripper import Gripper, GripperStage


_stop = False


def _on_signal(_signum, _frame):
    global _stop
    _stop = True


def main() -> int:
    default_port = "COM3" if sys.platform == "win32" else "/dev/ttyUSB0"
    port   = sys.argv[1] if len(sys.argv) > 1 else default_port
    cycles = int(sys.argv[2]) if len(sys.argv) > 2 else 5

    signal.signal(signal.SIGINT, _on_signal)

    try:
        with Gripper() as g:
            g.connect(port)

            if g.read_state().stage != GripperStage.ACTIVE:
                print("Activating ...")
                g.activate()
            if not g.read_state().homed:
                print("Homing ...")
                g.home()

            for i in range(cycles):
                if _stop:
                    break
                print(f"Cycle {i + 1}/{cycles}")

                print("  close ...")
                g.move_to(255, speed=200, force=128, blocking=True)
                print(f"    {g.read_state()}")

                if _stop:
                    break

                print("  open ...")
                g.move_to(0, speed=200, force=128, blocking=True)
                print(f"    {g.read_state()}")

            print("Interrupted." if _stop else "Done.")
    except Exception as e:
        print(f"ERROR: {e}", file=sys.stderr)
        return 1
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: cpp/examples/02_open_close_loop.cpp
// Example 2: Repeatedly open and close the gripper.
//
// Usage:
//   ./02_open_close_loop /dev/ttyUSB0 [cycles]
//   02_open_close_loop.exe COM3 [cycles]
#include "aidin/gripper.hpp"

#include <atomic>
#include <csignal>
#include <iostream>

namespace {
std::atomic<bool> g_stop{false};
void onSignal(int) { g_stop.store(true); }
}

int main(int argc, char** argv) {
    const std::string port = (argc > 1) ? argv[1]
#if defined(_WIN32)
                                        : "COM3";
#else
                                        : "/dev/ttyUSB0";
#endif
    const int cycles = (argc > 2) ? std::atoi(argv[2]) : 5;

    std::signal(SIGINT, onSignal);

    try {
        aidin::Gripper g;
        g.connect(port);

        if (g.readState().stage != aidin::GripperStage::Active) {
            std::cout << "Activating ...\n";
            g.activate();
        }
        if (!g.readState().homed) {
            std::cout << "Homing ...\n";
            g.home();
        }

        for (int i = 0; i < cycles && !g_stop.load(); ++i) {
            std::cout << "Cycle " << (i + 1) << "/" << cycles << "\n";

            std::cout << "  close ...\n";
            g.moveTo(/*pos=*/255, /*speed=*/200, /*force=*/128, /*blocking=*/true);
            std::cout << "    " << g.readState() << "\n";

            if (g_stop.load()) break;

            std::cout << "  open ...\n";
            g.moveTo(/*pos=*/0, /*speed=*/200, /*force=*/128, /*blocking=*/true);
            std::cout << "    " << g.readState() << "\n";
        }

        std::cout << (g_stop.load() ? "Interrupted.\n" : "Done.\n");
    } catch (const std::exception& e) {
        std::cerr << "ERROR: " << e.what() << "\n";
        return 1;
    }
    return 0;
}
